diesel/src/insertable.rs

```diff
@@ -36,33 +38,41 @@ pub enum ColumnInsertValue<Col, Expr> where
     Default(Col),
 }
 
+type ValuesFn<Item, T, DB> = fn(Item) -> <Item as Insertable<T, DB>>::Values;
+
 impl<Iter, T, DB> Insertable<T, DB> for Iter where
     T: Table,
     DB: Backend + SupportsDefaultKeyword,
     Iter: IntoIterator,
     Iter::Item: Insertable<T, DB>,
+    Iter::IntoIter: Clone,
 {
-    type Values = BatchInsertValues<<Iter::Item as Insertable<T, DB>>::Values>;
+    type Values = BatchInsertValues<iter::Map<
+        Iter::IntoIter,
+        ValuesFn<Iter::Item, T, DB>,
+    >>;
 
     fn values(self) -> Self::Values {
-        let values = self.into_iter().map(Insertable::values).collect();
+        let values = self.into_iter()
+            .map(Insertable::values as ValuesFn<Iter::Item, T, DB>);
```

Comment:
Nice. (Here was a longer comment musing about fn pointers. Shouldn't have read this commit-by-commit.)